core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -159,8 +162,27 @@ private[spark] class BlockManager(
     blockTransferService.init(this)
     shuffleClient.init(appId)
 
+    val topologyInfo = {
+      val topologyStr = master.getTopologyInfo(blockTransferService.hostName)
+      if (topologyStr == null || topologyStr.isEmpty) {
+        None
+      } else {
+        Some(topologyStr)
+      }
+    }
+
+    blockReplicationPrioritizer = {
+      val priorityClass = conf.get(
+        "spark.replication.topologyawareness.prioritizer",
+        "org.apache.spark.storage.DefaultBlockReplicationPrioritization")
+      val clazz = Utils.classForName(priorityClass)
+      val ret = clazz.newInstance.asInstanceOf[BlockReplicationPrioritization]
+      logInfo(s"Using $priorityClass for prioritizing peers")
+      ret
+    }
+
     blockManagerId = BlockManagerId(
-      executorId, blockTransferService.hostName, blockTransferService.port)
+      executorId, blockTransferService.hostName, blockTransferService.port, topologyInfo)
```

Comment:
Would it work if topologyInfo was sent back from the master when `registerBlockManager` is called? It doesn't seem that anything uses `blockManagerId` until registration finishes. That way we wouldn't need this weird two-step registration.